### PR TITLE
Panorama enhancement - add target arg

### PR DESCRIPTION
**Pull request by omerKarkKatz**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [ link to the issue](https://github.com/demisto/etc/issues/36402)

## Description
The panorama integration works on 2 kinds of configurations. the panorama configuration and a single FW configuration. 
In this PR we add the target argument to support invoking some commands with the panorama configuration on a single FW in panorama. 

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Review by GuyAfik — COMMENTED** [review]
Hi @omerKarkKatz 
Great work on that target add-on!

Please make sure you add unit-tests for the command which you changed to check that target was actually added to the query parameters if possible.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 1]
```diff
@@ -562,8 +562,9 @@ def panorama_commit(args):
     params = {
         'type': 'commit',
         'cmd': f'<commit>{command}</commit>',
-        'key': API_KEY
+        'key': API_KEY,
     }
+    update_target(args, params)
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:567]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 2]
```diff
@@ -2345,7 +2353,7 @@ def calculate_dbot_score(category: str, additional_suspicious: list, additional_
     return dbot_score
 
 
-def panorama_get_url_category_command(url_cmd: str, url: str, additional_suspicious: list, additional_malicious: list):
+def panorama_get_url_category_command(url_cmd: str, url: str, additional_suspicious: list, additional_malicious: list, target: str = None):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:2356]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 3]
```diff
@@ -2273,13 +2279,15 @@ def panorama_edit_custom_url_category_command(args: dict):
 
 
 @logger
-def panorama_get_url_category(url_cmd: str, url: str):
+def panorama_get_url_category(url_cmd: str, url: str, target: str):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:2282]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 4]
```diff
@@ -2794,17 +2802,39 @@ def prettify_rule(rule: dict):
...
+        if target:
+            if not target_filter(rule, target):
+                continue
         pretty_rule = prettify_rule(rule)
         pretty_rules_arr.append(pretty_rule)
 
     return pretty_rules_arr
 
 
+def target_filter(rule, target):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:2819]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 5]
```diff
@@ -2794,17 +2802,39 @@ def prettify_rule(rule: dict):
...
+    """
+        Args:
+            rule (dict): A rule from the panorama instance.
+            target (num): A serial number to filter the rule on
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:2823]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 6]
```diff
@@ -2794,17 +2802,39 @@ def prettify_rule(rule: dict):
...
+
+        Returns:
+            if the rule contains the target return True else False.
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:2826]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 7]
```diff
@@ -4136,14 +4168,14 @@ def panorama_query_traffic_logs_command(args: dict):
 
 
 @logger
-def panorama_get_traffic_logs(job_id: str):
+def panorama_get_traffic_logs(job_id: str, args: dict = {}):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:4171]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 8]
```diff
@@ -4136,14 +4168,14 @@ def panorama_query_traffic_logs_command(args: dict):
...
     params = {
         'action': 'get',
         'type': 'log',
         'job-id': job_id,
         'key': API_KEY
     }
-
+    update_target(args, params)
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:4178]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 9]
```diff
@@ -3692,6 +3694,9 @@ script:
     - description: The serial number for a virtual system commit. If provided, the
         commit will be a virtual system commit.
       name: serial_number
+      deprecated: true
+    - description: Use only with a Panorama configuration, invoke the command on the specific target - FW seriel number.
+      name: target
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3699]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 10]
```diff
@@ -3692,6 +3694,9 @@ script:
     - description: The serial number for a virtual system commit. If provided, the
         commit will be a virtual system commit.
       name: serial_number
+      deprecated: true
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3697]: Bat Katzir asked me to change the serial number arg and change it to target. he asked to deprecate the old one is this how its done?

File: Packs/PAN-OS/ReleaseNotes/1_11_3.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  - ***pan-os-content-update-download-status***
+  - ***pan-os-install-latest-content-update***
+  - ***pan-os-content-update-install-status***
+  - ***pan-os-check-latest-panos-software***
+  - ***pan-os-download-panos-version***
+  - ***pan-os-download-panos-status***
+  - ***pan-os-install-panos-version***
+  - ***pan-os-install-panos-status***
+
+- Added a human-readable for the ***panorama-check-latest-panos-software*** to present top 5 most recent software releases.
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/ReleaseNotes/1_11_3.md:26]: no it returned the the output as a dict

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 12]
```diff
@@ -4413,14 +4446,15 @@ def panorama_query_logs_command(args: dict):
     rule = args.get('rule')
     filedigest = args.get('filedigest')
     url = args.get('url')
+    target = args.get('target', None)
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:4457]: - Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 13]
```diff
@@ -4448,13 +4482,14 @@ def panorama_query_logs_command(args: dict):
     })
 
 
-def panorama_check_logs_status_command(job_id: str):
+def panorama_check_logs_status_command(args: dict):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:4493]: Talked to you about this

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 14]
```diff
@@ -2838,8 +2868,10 @@ def panorama_list_rules_command(tag: str):
     else:
         xpath = XPATH_SECURITY_RULES
 
+    tag = args.get('tag')
+    target = args.get('target', None)
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:2878]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 15]
```diff
@@ -6197,12 +6215,12 @@ script:
       description: Content installation status details.
       type: String
   - arguments:
-    - description: The target device from which to get the PAN-OS software version.
+    - description: Use only with a Panorama configuration, invoke the command on the specific target - FW seriel number.
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:6230]: on purpose i maintained a uniformed text for all target args

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 16]
```diff
@@ -756,11 +762,11 @@ def safeget(dct: dict, keys: List[str]):
     return dct
 
 
-def panorama_push_status_command(job_id: str):
+def panorama_push_status_command(args: dict):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:768]: wanted it to be as uniform as possible

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3698]: took a closer look and decided to remove this code.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 17]
```diff
@@ -564,6 +564,8 @@ def panorama_commit(args):
         'cmd': f'<commit>{command}</commit>',
         'key': API_KEY
     }
+    if target := args.get('target', None):
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:567]:
```suggestion
    if target := args.get('target'):
```

get method of a dict returns ```None``` by default.
not super important change though, the way you did is good too :)

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 18]
```diff
@@ -2794,17 +2808,39 @@ def prettify_rule(rule: dict):
...
 
 
+def target_filter(rule: dict, target: str):
+    """
+    Args:
+        rule (dict): A rule from the panorama instance.
+        target (str): A serial number to filter the rule on
+
+    Returns:
+        if the rule contains the target return True else False.
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:2832]:
```suggestion
        True if the rule contains the firewall serial number (target), False if not.
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 19]
```diff
@@ -5430,7 +5472,15 @@ def panorama_check_latest_panos_software_command(target: Optional[str] = None):
         'GET',
         params=params
     )
-    return_results(result['response']['result'])
+    to_context = result.get('response', {}).get('result', {})
+    versions = to_context.get('sw-updates', {}).get('versions').get('entry', [])
+    if len(versions) > 5:
+        versions = versions[:5]
+    human_readable = tableToMarkdown('5 latest pan-os software releases', versions, ['version', 'filename', 'size', 'released-on', 'downloaded' , 'current' , 'latest', 'uploaded'] )
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:5479]:
```suggestion
    human_readable = tableToMarkdown('5 latest pan-os software releases', versions, ['version', 'filename', 'size', 'released-on', 'downloaded' , 'current' , 'latest', 'uploaded'], removeNull=True)
```

File: Packs/PAN-OS/ReleaseNotes/1_11_3.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (GuyAfik) [reply Packs/PAN-OS/ReleaseNotes/1_11_3.md:25]: i don't think that one is necessary, please remove it.

File: Packs/PAN-OS/ReleaseNotes/1_11_3.md [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+  - ***pan-os-show-device-version***
+  - ***pan-os-download-latest-content-update***
+  - ***pan-os-content-update-download-status***
+  - ***pan-os-install-latest-content-update***
+  - ***pan-os-content-update-install-status***
+  - ***pan-os-check-latest-panos-software***
+  - ***pan-os-download-panos-version***
+  - ***pan-os-download-panos-status***
+  - ***pan-os-install-panos-version***
+  - ***pan-os-install-panos-status***
```
Comment (GuyAfik) [inline Packs/PAN-OS/ReleaseNotes/1_11_3.md:23]:
```suggestion
#### Integrations
##### Palo Alto Networks PAN-OS
- Added support for the *target* argument in the following commands.  
  - ***pan-os-get-url-category***
  - ***pan-os-device-reboot***
  - ***pan-os-commit***
  - ***pan-os-commit-status***
  - ***pan-os-push-status***
  - ***pan-os-list-rules***
  - ***pan-os-query-logs***
  - ***pan-os-check-logs-status***
  - ***pan-os-get-logs***
  - ***pan-os-show-device-version***
  - ***pan-os-download-latest-content-update***
  - ***pan-os-content-update-download-status***
  - ***pan-os-install-latest-content-update***
  - ***pan-os-content-update-install-status***
  - ***pan-os-check-latest-panos-software***
  - ***pan-os-download-panos-version***
  - ***pan-os-download-panos-status***
  - ***pan-os-install-panos-version***
  - ***pan-os-install-panos-status***
- Documentation and metadata improvements.
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 21]
```diff
@@ -5430,7 +5472,15 @@ def panorama_check_latest_panos_software_command(target: Optional[str] = None):
...
+    return CommandResults(readable_output=human_readable,
+                                  outputs=to_context,
+                                  raw_response=result,
+                                  )
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:5483]:
```suggestion
    return CommandResults(readable_output=human_readable,
                                  outputs=to_context,
                                  raw_response=result,
                                  output_prefix=?
                                  outputs_key_field=?
                                  )
```
discuss what should be the outputs prefix with @bakatzir if needed.

**Review by GuyAfik — COMMENTED** [review]
@omerKarkKatz nice work buddy!

Can you please also add the **target** argument to the relevant commands in the README?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 22]
```diff
@@ -3581,7 +3581,7 @@ script:
     - description: The time that the PCAP was received on the firewall. Used for threat
         PCAPs.
       name: search-time
-    - description: Target number of the firewall. Use only on a Panorama instance.
+    - description:  Use only with a Panorama instance. Serial number of the firewall on which to run the command.
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3584]: I dont find it redundant will take it to Bar Katzir

[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3584]: @bakatzir

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 21 above)
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Done

**Comment by ShirleyDenkberg** [comment]
@GuyAfik Doc review completed.

File: Packs/PAN-OS/pack_metadata.json [hunk 23]
```diff
@@ -2,7 +2,7 @@
     "name": "PAN-OS",
     "description": "Manage Palo Alto Networks Firewall and Panorama. For more information see Panorama documentation.",
     "support": "xsoar",
-    "currentVersion": "1.12.1",
+    "currentVersion": "1.12.2",
```
Comment (yaakovi) [inline Packs/PAN-OS/pack_metadata.json:5]: bump by minor (1.13.0)

File: Packs/PAN-OS/ReleaseNotes/1_12_2.md [hunk 24]
```diff
@@ -0,0 +1,23 @@
+
+#### Integrations
+##### Palo Alto Networks PAN-OS- Added support for the *target* argument in the following commands.
```
Comment (yaakovi) [inline Packs/PAN-OS/ReleaseNotes/1_12_2.md:3]:
```suggestion
##### Palo Alto Networks PAN-OS
- Added support for the *target* argument in the following commands.  
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 25]
```diff
@@ -612,6 +614,10 @@ def panorama_commit_status(args: dict):
         'cmd': f'<show><jobs><id>{args.get("job_id")}</id></jobs></show>',
         'key': API_KEY
     }
+
+    if target := args.get('target', None):
```
Comment (yaakovi) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:618]:
```suggestion
    if target := args.get('target'):
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 26]
```diff
@@ -2795,17 +2809,39 @@ def prettify_rule(rule: dict):
...
-def prettify_rules(rules: Union[List[dict], dict]):
+def prettify_rules(rules: Union[List[dict], dict], target: Optional[str] = None):
     if not isinstance(rules, list):
-        return prettify_rule(rules)
+        rules = [rules]
     pretty_rules_arr = []
     for rule in rules:
+        if target:
+            if not target_filter(rule, target):
+                continue
```
Comment (yaakovi) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:2819]:
```suggestion
        if target and not target_filter(rule, target):
            continue
```

**Review by yaakovi — COMMENTED** [review]
looks good

**Review by yaakovi — APPROVED** [review]

File: Packs/PAN-OS/pack_metadata.json (hunk 23 above)
Comment (omerKarkKatz) [reply Packs/PAN-OS/pack_metadata.json]: Done

File: Packs/PAN-OS/Integrations/Panorama/README.md [hunk 27]
```diff
@@ -3624,10 +3626,10 @@ Installs the target PAN-OS version on the specified target device.
 `pan-os-install-panos-version`
 #### Input
 
-| **Argument Name** | **Description** | **Required** |
-| --- | --- | --- |
-| target | The target device on which to install the target PAN-OS software version. | Optional | 
-| target_version | Target PAN-OS version to install. | Required | 
+| **Argument Name** | **Description**                                                                              | **Required** |
+| --- |----------------------------------------------------------------------------------------------| --- |
```
Comment (yaakovi) [inline Packs/PAN-OS/Integrations/Panorama/README.md:3630]: Why?

**Review by yaakovi — APPROVED** [review]

File: Packs/PAN-OS/Integrations/Panorama/README.md (hunk 27 above)
Comment (omerKarkKatz) [reply Packs/PAN-OS/Integrations/Panorama/README.md:3630]: fixed it i guess it happens automaticlly with generate docs

**Comment by bakatzir** [comment]
@ShirleyDenkberg would you mind re-review the RN and readme changes?

File: Packs/PAN-OS/Integrations/Panorama/README.md [hunk 28]
```diff
@@ -48,6 +48,7 @@ This integration was integrated and tested with version 8.1.0 and 9.0.1 of Palo
    * [pan-os-query-logs](#pan-os-query-logs)
    * [pan-os-check-logs-status](#pan-os-check-logs-status)
    * [pan-os-get-logs](#pan-os-get-logs)
+* The target argument works only on operational type commands. (not commit, logs, etc)
```
Comment (bakatzir) [inline Packs/PAN-OS/Integrations/Panorama/README.md:51]:
```suggestion
* The target argument works only on operational type commands. Meaning, you can not use it with commit, logs, PCAP commands.
```

File: Packs/PAN-OS/ReleaseNotes/1_14_3.md [hunk 29]
```diff
@@ -0,0 +1,19 @@
+
+#### Integrations
+##### Palo Alto Networks PAN-OS
+- Added support for the *target* argument in the following commands. The target argument is supported only by panorama op type commands. (not including logs, commits, etc)
```
Comment (bakatzir) [inline Packs/PAN-OS/ReleaseNotes/1_14_3.md:4]:
```suggestion
- Added support for the *target* argument in the following commands. The target argument is supported only in operational type commands. Meaning, you can not use it with commit, logs, PCAP commands.
```

File: Packs/PAN-OS/ReleaseNotes/1_14_3.md [hunk 30]
```diff
@@ -0,0 +1,19 @@
...
+  - ***pan-os-content-update-download-status***
+  - ***pan-os-install-latest-content-update***
+  - ***pan-os-content-update-install-status***
+  - ***pan-os-check-latest-panos-software***
+  - ***pan-os-download-panos-version***
+  - ***pan-os-download-panos-status***
+  - ***pan-os-install-panos-version***
+  - ***pan-os-install-panos-status***
+  - ***pan-os***
+- Added support for human-readable output for the ***pan-os-check-latest-panos-software*** command to present the top 5 most recent software releases.
```
Comment (bakatzir) [inline Packs/PAN-OS/ReleaseNotes/1_14_3.md:19]:
```suggestion
- Added in the ***pan-os-check-latest-panos-software*** command the top 5 most recent software releases.
```

**Comment by ShirleyDenkberg** [comment]
@bakatzir @yaakovi @GuyAfik Doc review completed.

**Pull request merged**